doc/manual/src/contributing/json-guideline.md

```diff
@@ -0,0 +1,132 @@
+
+This leads to the following guidelines:
+
+ - The top-level (root) value must be a record.
+
+   Otherwise, one can not change the structure of a command's output.
+
+ - The value of a dictionary item must be a record.
+
+   Otherwise, the item type can not be extended.
+
+ - List items should be records.
+
+   Otherwise, one can not change the structure of the list items.
+
+   If the order of the items does not matter, and each item has a unique key that is a string, consider representing the list as a dictionary instead.
+   If the order of the items needs to be preserved, return a list of records.
+
+ - Streaming JSON should return records.
+
+   An example of a streaming JSON format is [JSON lines](https://jsonlines.org/), where each line represents a JSON value.
+   These JSON values can be considered top-level values or list items, and they must be records.
+
+### Examples
+
+This is bad, because all keys must be assumed to be store types:
+
+```json
+{
+  "local": { ... },
+  "remote": { ... },
+  "http": { ... }
+}
+```
+
+This is good, because the it is extensible at the root, and is somewhat self-documenting:
+
+```json
+{
+  "storeTypes": { "local": { ... }, ... },
+  "pluginSupport": true
+}
+```
+
+While the dictionary of store types seems like a very complete response at first, a use case may arise that warrants returning additional information.
+For example, the presence of plugin support may be crucial information for a client to proceed when their desired store type is missing.
+
+
+
+The following representation is bad because it is not extensible:
+
+```json
+{ "outputs": [ "out" "bin" ] }
+```
+
+However, simply converting everything to records is not enough, because the order of outputs must be preserved:
+
+```json
+{ "outputs": { "bin": {}, "out": {} } }
+```
+
+The first item is the default output. Deriving this information from the outputs ordering is not great, but this is how Nix currently happens to work.
+While it is possible for a JSON parser to preserve the order of fields, we can not rely on this capability to be present in all JSON libraries.
+
+This representation is extensible and preserves the ordering:
+
+```json
+{ "outputs": [ { "outputName": "out" }, { "outputName": "bin" } ] }
+```
+
+## Ease of consumption
+
+Per the above, its crucial we should be able to add fields to extend the format in a backwards-compatible way.
+However, that should *not* mean we use the presence/absence of fields to indicate optional information *within* a version of the format.
+Instead, one should always include the field, and use `null` to indicate the "nothing" case.
+
+An exception to this is [tagged unions](https://en.wikipedia.org/wiki/Tagged_union).
+In this case, we have one "tag" field which says what *variant* the record should be, and a bunch of other variant-specific fields.
+There is no expectation to include `null`-valued fields from other variants.
```

Comment:
This needs an example, but I don't think we've formalized how we represent tagged unions yet.
There's multiple ways; see e.g. [`attrTag`](https://nixos.org/manual/nixos/stable/#sec-option-types-sums) vs [`taggedSubmodule`](https://github.com/NixOS/nixpkgs/pull/254790). Another one is a two-field records where the value field's schema is dependent on the tag field's value.

Postpone to follow-up PR?